ichnaea/locate/provider.py

```diff
@@ -531,3 +545,63 @@ class GeoIPPositionProvider(BaseGeoIPProvider):
 
 class GeoIPCountryProvider(BaseGeoIPProvider):
     location_type = Country
+
+
+class FallbackProvider(Provider):
+    log_name = 'fallback'
+    location_type = Position
+    source = DataSource.Fallback
+
+    @property
+    def fallback_url(self):
+        return read_config().get_map('ichnaea')['fallback_url']
+
+    def should_locate(self, data, location):
+        empty_location = not location.found()
+        weak_location = location.source >= DataSource.GeoIP
+        cell_found = data.get('cell', None)
+        wifi_found = len(data.get('wifi', [])) > 1
+        return (
+            self.api_key.allow_fallback and
+            (empty_location or weak_location) and
+            (cell_found or wifi_found)
+        )
+
+    def locate(self, data):
+        location = self.location_type(query_data=False)
+
+        cell_queries = [{
+            'cellId': cell.get('cid', None),
+            'locationAreaCode': cell.get('lac', None),
+            'mobileCountryCode': cell.get('mcc', None),
+            'mobileNetworkCode': cell.get('mnc', None),
+        } for cell in data.get('cell', [])]
+        wifi_queries = [{
+            'macAddress': re.sub(
+                '(..(?!$))', '\g<0>:', wifi.get('key', None)),
+        } for wifi in data.get('wifi', [])]
+
+        query_data = {
+            'cellTowers': cell_queries,
+            'wifiAccessPoints': wifi_queries,
+        }
+
+        try:
+            response = requests.post(
+                self.fallback_url, json=query_data)
+        except requests.exceptions.RequestException:
```

Comment:
Can you add an override for the user-agent header? So it becomes something like I did in the export code at https://github.com/mozilla/ichnaea/blob/master/ichnaea/data/export.py#L105

Something like user-agent: ichnaea, so we don't leak what specific version of urllib/python we are using internally.

I'd also add the other two things I did there: 1. add a timeout, for this lookup make it 5 seconds and disable ssl verification for now.

IIRC we have to do a bunch of work to get SNI support working in Python 2.6.